flink-ml-core/src/main/java/org/apache/flink/ml/param/ArrayArrayParam.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.param;
+
+import org.apache.flink.ml.util.ReadWriteUtils;
+
+import java.io.IOException;
+
+/** Class for the arrayOfArray-type parameters. */
+public class ArrayArrayParam<T> extends Param<T[][]> {
```

Comment:
We probably follow the user behavior here. 

In Bucketizer, users specify a split array for each input col. Moreover, the number of split points in different columns could be different. If we use 1D array to represent 2D array, we would also need to specify the dimension for each 1D array. This seems a bit complex to me.

Also, using 2D array is consistent with exisiting libraries [1].

[1] https://github.com/apache/spark/blob/1018f8b545a74f4fcaf123b5f15d52ac2afa2d3c/mllib/src/main/scala/org/apache/spark/ml/param/params.scala#L567

Reply:
I agree with it that on API we need to display getter/setter for 2D arrays, but as for internal implementation, I believe we can store the 2D array in a 1D format. This does not affect user experience and can reuse existing `ArrayParams`.

Reply:
> we can store the 2D array in a 1D format

I am a bit confused about this. How are we going to store the shape of each dimension? 
Maintaining another `array` for shapes seems a bit complex to me.

Reply:
How about this?
```java
    DoubleArrayParam SPLITS_ARRAY =
            new DoubleArrayParam(
                    "splitsArray", "Array of split points.", null, ParamValidators.nonEmptyArray());

    default Double[][] getSplitsArray() {
        Double[] values = get(SPLITS_ARRAY);
        Double[][] result = new Double[values[0].intValue()][];
        int offset = 1;
        for (int i = 0; i < values[0].intValue(); i++) {
            result[i] = new Double[values[offset].intValue()];
            System.arraycopy(values, offset + 1, result[i], 0, values[offset].intValue());
            offset += values[0].intValue() + 1;
        }
        return result;
    }

    default T setSplitsArray(Double[][] value) {
        List<Double> list = new ArrayList<>();
        list.add((double) value.length);
        for (Double[] doubles : value) {
            list.add((double) doubles.length);
            list.addAll(Arrays.asList(doubles));
        }
        set(SPLITS_ARRAY, list.toArray(new Double[0]));
        return (T) this;
    }
```

My point is that we should try to simplify the internal implementation. Currently we have about 6 basic type params, including String, Boolean, Int, Long, Double, Float. For array-typed parameters, we introduced another 6 classes. For 2D-array typed parameters, we would also need 6 classes. Same for 3D, 4D cases. I don't think it would be a good idea to have so many classes for parameters, each of which shares a large proportion in common. For this problem, the solution I proposed in comments above is to reuse 1D classes for arrays of larger dimensions.

Reply:
I am not sure the alternative approach simplify the overall internal implementation.

The alternative approach does remove the need for DoubleArrayArrayParam. But for every parameter of type `double[][]`, it requires the setXXX() and getXXX() to have non-trivial implementation as describe above. We can add two public static methods to be shared by those setXXX() and getXXX(). But the addition of these two methods do not seem to be simpler than the existing `DoubleArrayArrayParam`.